exporter/trace/otelcol/transform_spans.go

```diff
@@ -0,0 +1,157 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otelcol
+
+import (
+	"google.golang.org/grpc/codes"
+
+	commonpb "github.com/open-telemetry/opentelemetry-proto/gen/go/common/v1"
+	tracepb "github.com/open-telemetry/opentelemetry-proto/gen/go/trace/v1"
+
+	"go.opentelemetry.io/otel/api/core"
+	apitrace "go.opentelemetry.io/otel/api/trace"
+	export "go.opentelemetry.io/otel/sdk/export/trace"
+)
+
+const (
+	maxMessageEventsPerSpan = 128
+)
+
+func otSpanToProtoSpan(sd *export.SpanData) *tracepb.Span {
+	if sd == nil {
+		return nil
+	}
+	return &tracepb.Span{
+		TraceId:           sd.SpanContext.TraceID[:],
+		SpanId:            sd.SpanContext.SpanID[:],
+		ParentSpanId:      sd.ParentSpanID[:],
+		Status:            otStatusToProtoStatus(sd.Status),
+		StartTimeUnixnano: uint64(sd.StartTime.Nanosecond()),
+		EndTimeUnixnano:   uint64(sd.EndTime.Nanosecond()),
+		Links:             otLinksToProtoLinks(sd.Links),
+		Kind:              otSpanKindToProtoSpanKind(sd.SpanKind),
+		Name:              sd.Name,
+		Attributes:        otAttributesToProtoAttributes(sd.Attributes),
+		Events:            otTimeEventsToProtoTimeEvents(sd.MessageEvents),
+		// TODO (rghetia): Add Tracestate: when supported.
+		DroppedAttributesCount: uint32(sd.DroppedAttributeCount),
+		DroppedEventsCount:     uint32(sd.DroppedMessageEventCount),
+		DroppedLinksCount:      uint32(sd.DroppedLinkCount),
+	}
+}
+
+func otStatusToProtoStatus(status codes.Code) *tracepb.Status {
+	return &tracepb.Status{
+		Code: tracepb.Status_StatusCode(status),
+		// TODO (rghetia) : Add Status Message: when supported.
+	}
+}
+
+func otLinksToProtoLinks(links []apitrace.Link) []*tracepb.Span_Link {
+	if len(links) == 0 {
+		return nil
+	}
+
+	sl := make([]*tracepb.Span_Link, 0, len(links))
+	for _, otLink := range links {
+		// This redefinition is necessary to prevent otLink.*ID[:] copies
+		// being reused -- in short we need a new otLink per iteration.
+		otLink := otLink
+
+		sl = append(sl, &tracepb.Span_Link{
+			TraceId:    otLink.TraceID[:],
+			SpanId:     otLink.SpanID[:],
+			Attributes: otAttributesToProtoAttributes(otLink.Attributes),
+		})
+	}
+	return sl
+}
+
+func otAttributesToProtoAttributes(attrs []core.KeyValue) []*commonpb.AttributeKeyValue {
+	if len(attrs) == 0 {
+		return nil
+	}
+	out := make([]*commonpb.AttributeKeyValue, 0, len(attrs))
+	for _, v := range attrs {
+		switch v.Value.Type() {
+		case core.BOOL:
+			out = append(out, &commonpb.AttributeKeyValue{
+				Key:       string(v.Key),
+				Type:      commonpb.AttributeKeyValue_BOOL,
+				BoolValue: v.Value.AsBool(),
+			})
+		case core.INT64, core.INT32, core.UINT32, core.UINT64:
+			out = append(out, &commonpb.AttributeKeyValue{
+				Key:      string(v.Key),
+				Type:     commonpb.AttributeKeyValue_INT,
+				IntValue: v.Value.AsInt64(),
```

Comment:
This looks OK to me, because when the Value is constructed for a int32 value, we call
```
func int32ToRaw(i int32) uint64 {
	return uint64(i)
}
```